File: docs/antikythera-maths/research-mfo/fractal_sweep.py
import numpy as np
import json
from pathlib import Path
import heapq

# Standard Model target masses in GeV
sm_masses_gev = {
    'electron': 0.000511,
    'up': 0.0022,
    'down': 0.0047,
    'strange': 0.095,
    'muon': 0.1057,
    'charm': 1.275,
    'tau': 1.777,
    'bottom': 4.18,
    'top': 173.0,
}

# Target: ratios of mass squared relative to electron, sorted
m_e = sm_masses_gev['electron']
sm_mass_sq_ratios = sorted([(v/m_e)**2 for v in sm_masses_gev.values()])
target_log = np.log10(sm_mass_sq_ratios)

def generalized_decimation(n, L, max_level=5):
    """
    Generalized spectral decimation for a fractal with n copies
    and decimation factor L.
    """
    def R_inverse(w):
        disc = L*L - 4*w
        if disc < 0:
            return []
        sd = np.sqrt(disc)
        return [(L - sd)/2.0, (L + sd)/2.0]

    memo = {}
    def get_evals_at_level(m):
        if m == 0:
            return {0.0, float(L)}
        if m in memo:
            return memo[m]
        prev = get_evals_at_level(m - 1)
        new_evals = set()
        for w in prev:
            for z in R_inverse(w):
                if 0 <= z <= L + 1:
                    new_evals.add(round(z, 12))
        new_evals.add(2.0)
        new_evals.add(float(n + 2))
        memo[m] = new_evals
        return new_evals

    continuous_evals = set()
    for m in range(max_level + 1):
        for e in get_evals_at_level(m):
            if e > 1e-10:
                continuous_evals.add(round((L**m) * e, 6))
    return sorted(list(continuous_evals))

# Precompute fractal eigenvalues for each (n, L)
print("Precomputing fractal spectra...")
fractal_cache = {}
for n in range(3, 9):
    for L in range(4, 21):
        # We only need the smallest few to find the overall smallest 9 total
        evals = generalized_decimation(n, L, max_level=5)
        fractal_cache[(n, L)] = evals[:15]  # Keep top 15

# CP2 and S1 constants
cp2_base = [0, 12, 32, 60, 96, 140]
s1_n = [0, 1, 2, 3, 4]

# R sweep values: log steps from 0.1 to 10
r_values = np.logspace(-1, 1, 9)

candidates = []

print("Starting anisotropic product sweep...")
# To reduce redundant permutations (L1, L2, L3), we enforce L1 <= L2 <= L3
n_total = 6 * (17 * 18 * 19 // 6) * 9  # Approx count
count = 0

for n in range(3, 9):
    for L1 in range(4, 21):
        E1 = fractal_cache[(n, L1)]
        for L2 in range(L1, 21):
            E2 = fractal_cache[(n, L2)]
            for L3 in range(L2, 21):
                E3 = fractal_cache[(n, L3)]
                
                # Combine fractal axes
                fractal_sums = set()
                for e1 in E1:
                    for e2 in E2:
                        for e3 in E3:
                            fractal_sums.add(e1 + e2 + e3)
                # Add 0 to fractal sums to allow zero contributions from fractal axes
                # (Technically each axis has a 0 mode, but generalized_decimation returns non-zero)
                # In a product manifold, the 0 mode is always present.
                # Let's add it explicitly for the combined set.
                fractal_sums.add(0.0)
                sorted_fractal = sorted(list(fractal_sums))[:30]

                for R in r_values:
                    s1_evals = [(n_s**2) / (R**2) for n_s in s1_n]
                    
                    # Total combinations
                    combined = set()
                    for f in sorted_fractal:
                        for cp in cp2_base:
                            for s1 in s1_evals:
                                val = f + cp + s1
                                if val > 1e-10:
                                    combined.add(round(val, 6))
                    
                    if len(combined) < 9:
                        continue
                    
                    sorted_combined = sorted(list(combined))
                    top_9 = np.array(sorted_combined[:9])
                    normalized_9 = top_9 / top_9[0]
                    log_9 = np.log10(normalized_9)
                    
                    error = np.sum((log_9 - target_log)**2)
                    
                    candidates.append({
                        "n": n,
                        "L1": L1, "L2": L2, "L3": L3,
                        "R": float(R),
                        "error": float(error),
                        "ratios": normalized_9.tolist()
                    })
                    
                    count += 1
                    if count % 5000 == 0:
                        print(f"Processed {count} configurations...")

# Sort and save
candidates.sort(key=lambda x: x["error"])
top_100 = candidates[:100]

results_path = Path("results-mfo/anisotropic_product_sweep.json")
with open(results_path, "w") as f:
    json.dump(top_100, f, indent=2)

print("\nSweep complete.")
print(f"Top 10 candidates (sorted by log10 squared error):")
print(f"{'n':>2s} {'L1':>3s} {'L2':>3s} {'L3':>3s} {'R':>6s} {'Error':>10s}")
for c in candidates[:10]:
    print(f"{c['n']:2d} {c['L1']:3d} {c['L2']:3d} {c['L3']:3d} {c['R']:6.3f} {c['error']:10.4f}")

print(f"\nResults saved to {results_path}")

File: docs/antikythera-maths/research-mfo/fractal_sweep.jl
using JSON
using Printf
using LinearAlgebra
using Random

# Standard Model target masses in GeV
const SM_MASSES_GEV = [0.000511, 0.0022, 0.0047, 0.095, 0.1057, 1.275, 1.777, 4.18, 173.0]
const M_E = 0.000511
const TARGET_LOG = log10.(sort([(v/M_E)^2 for v in SM_MASSES_GEV]))

# Precomputed Manifold Constants
const CP2_BASE = Float64[0, 12, 32, 60, 96, 140]
const S1_N = Float64[0, 1, 4, 9, 16] 

function r_inverse(w, L)
    disc = L*L - 4*w
    if disc < 0 return Float64[] end
    sd = sqrt(disc)
    return [(L - sd)/2.0, (L + sd)/2.0]
end

function get_fractal_evals(n, L, max_level=5)
    res = Float64[]
    current_evals = Set{Float64}([0.0, Float64(L)])
    for m in 0:max_level
        factor = Float64(L)^m
        for e in current_evals
            if e > 1e-10
                push!(res, round(e * factor, digits=6))
            end
        end
        if m < max_level
            next_evals = Set{Float64}()
            for w in current_evals
                for z in r_inverse(w, L)
                    if 0 <= z <= L + 1
                        push!(next_evals, round(z, digits=12))
                    end
                end
            end
            push!(next_evals, 2.0)
            push!(next_evals, Float64(n + 2))
            current_evals = next_evals
        end
    end
    return sort(unique(res))[1:min(15, length(res))]
end

function run_sweep()
    Random.seed!(42) # Ensure reproducibility for any sampling
    n = 3
    l_range = 10:10:100
    C = 0.1 # Coupling scale
    
    println("Precomputing fractal spectra for n=3, L=10:10:100...")
    fractal_cache = Dict{Int, Vector{Float64}}()
    for L in l_range
        fractal_cache[L] = get_fractal_evals(n, L)
    end

    r_values = 10.0 .^ range(2, 4, step=0.5)
    candidates = []
    
    println("Starting direct product coupling sweep...")
    for L1 in l_range
        E1 = fractal_cache[L1]
        for L2 in filter(x -> x <= L1, l_range)
            E2 = fractal_cache[L2]
            for L3 in filter(x -> x <= L2, l_range)
                E3 = fractal_cache[L3]
                
                # Direct Product Coupling: C * e1 * e2 * e3
                # We interpret "randomly selected pairs" as a stochastic combination
                # for the combined fractal contribution.
                f_combos = Float64[0.0]
                
                # All combinations (approx 15^3 = 3375)
                for e1 in E1, e2 in E2, e3 in E3
                    push!(f_combos, C * e1 * e2 * e3)
                end
                
                unique_f = sort(unique(round.(f_combos, digits=6)))
                unique_f_top = unique_f[1:min(40, length(unique_f))]

                for R in r_values
                    r2_inv = 1.0 / (R*R)
                    
                    res_combined = Float64[]
                    for f in unique_f_top, cp in CP2_BASE, s1_n2 in S1_N
                        val = f + cp + (s1_n2 * r2_inv)
                        if val > 1e-10
                            push!(res_combined, round(val, digits=6))
                        end
                    end
                    
                    if length(res_combined) < 9 continue end
                    
                    final_9_unique = sort(unique(res_combined))[1:min(9, length(res_combined))]
                    if length(final_9_unique) < 9 continue end
                    
                    normalized_9 = final_9_unique ./ final_9_unique[1]
                    log_9 = log10.(normalized_9)
                    error = sum((log_9 .- TARGET_LOG).^2)
                    
                    push!(candidates, (n=n, L1=L1, L2=L2, L3=L3, R=R, error=error, ratios=normalized_9))
                end
            end
        end
    end

    sort!(candidates, by=x->x.error)
    top_20 = candidates[1:min(20, length(candidates))]
    
    results_path = "results-mfo/product_coupling_sweep.json"
    mkpath("results-mfo")
    open(results_path, "w") do f
        JSON.print(f, [Dict("n"=>c.n, "L1"=>c.L1, "L2"=>c.L2, "L3"=>c.L3, "R"=>c.R, "error"=>c.error, "ratios"=>c.ratios) for c in top_20], 2)
    end

    println("\nProduct coupling sweep complete. Results saved to $results_path")
    @printf("%3s %3s %3s %9s %10s\n", "L1", "L2", "L3", "R", "Error")
    for c in top_20[1:min(10, length(top_20))]
        @printf("%3d %3d %3d %9.3f %10.4f\n", c.L1, c.L2, c.L3, c.R, c.error)
    end
end

run_sweep()
